### PR TITLE
crypto/aspeed: ast2600: Add HACE HW hash support

### PR DESCRIPTION
**Pull request by ChiaweiW**
Aspeed AST2600 Hash and Crypto Engine (HACE) is designated to accelerate the throughput of hash and symmetric encryption/decryption.

The change in this PR adds the driver support for the hash part of HACE to provide HW-assisted hashing in TEE.
In addition, it also constructs the initial crypto driver structure for Aspeed crypto engines.

### CODE REVIEW

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+static TEE_Result ast2600_hace_process(struct crypto_hash_ctx *ctx,
+				       const uint8_t *data, size_t len)
+{
+	TEE_Result rc = TEE_SUCCESS;
+	uint32_t sts;
+	uint64_t tref;
+	struct ast2600_hace_ctx *hace_ctx;
+
+	hace_ctx = (struct ast2600_hace_ctx *)(void *)ctx;
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:90]: I think the proper way is to use `container_of()`, see `core/drivers/crypto/caam/hash/caam_hash.c`.

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+static void ast2600_hace_copy_state(struct crypto_hash_ctx *dst_ctx,
+				    struct crypto_hash_ctx *src_ctx)
+{
+	cache_operation(TEE_CACHEINVALIDATE, src_ctx,
+			sizeof(struct ast2600_hace_ctx));
+
+	memcpy(dst_ctx, src_ctx, sizeof(struct ast2600_hace_ctx));
+
+	cache_operation(TEE_CACHEFLUSH, dst_ctx,
+			sizeof(struct ast2600_hace_ctx));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:289]: Align with `TEE_CACHEFLUSH`

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+	*pctx = (struct crypto_hash_ctx *)hace_ctx;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ast2600_drvcrypt_register_hash(void)
+{
+	hace_virt = core_mmu_get_va(HACE_BASE,
+				    MEM_AREA_IO_SEC, SMALL_PAGE_SIZE);
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:351]: `MEM_AREA_IO_SEC,` fits on the previous line

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+	if (!hace_virt) {
+		EMSG("cannot get HACE virtual address\n");
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:353]: `\n` not needed

File: core/drivers/crypto/aspeed/crypto_ast2600.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	io_write32(scu_virt + SCU_RST1, SCU_RST_CRYPTO);
+	udelay(100);
+	io_write32(scu_virt + SCU_RSTCLR1, SCU_RST_CRYPTO);
+
+	if (IS_ENABLED(CFG_CRYPTO_DRV_HASH)) {
+		io_write32(scu_virt + SCU_CLKGATECLR1, SCU_CLKGATE_HACE);
+
+		rc = ast2600_drvcrypt_register_hash();
+		if (rc) {
+			EMSG("cannot register hash driver, rc=%d\n", rc);
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/crypto_ast2600.c:43]: `\n` not needed

File: core/drivers/crypto/aspeed/crypto_ast2600.c [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#define SCU_CLKGATECLR1		0x84
+#define SCU_CLKGATE_HACE	BIT(13)
+
+static TEE_Result crypto_ast2600_init(void)
+{
+	TEE_Result rc = TEE_SUCCESS;
+	vaddr_t scu_virt = 0;
+
+	scu_virt = core_mmu_get_va(SCU_BASE,
+				   MEM_AREA_IO_NSEC, SMALL_PAGE_SIZE);
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/crypto_ast2600.c:29]: `MEM_AREA_IO_NSEC,` fits on previous line

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+static void ast2600_hace_free(struct crypto_hash_ctx *ctx)
+{
+	free(ctx);
+}
+
+static void ast2600_hace_copy_state(struct crypto_hash_ctx *dst_ctx,
+				    struct crypto_hash_ctx *src_ctx)
+{
+	cache_operation(TEE_CACHEINVALIDATE, src_ctx,
+			sizeof(struct ast2600_hace_ctx));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:284]: Align with `TEE_CACHEINVALIDATE`

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+static const struct crypto_hash_ops ast2600_hace_ops = {
+	.init = ast2600_hace_init,
+	.update = ast2600_hace_update,
+	.final = ast2600_hace_final,
+	.free_ctx = ast2600_hace_free,
+	.copy_state = ast2600_hace_copy_state,
+};
+
+static TEE_Result ast2600_hace_alloc(struct crypto_hash_ctx **pctx,
+				     uint32_t algo)
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:301]: Alignment

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be_h,
+					 sizeof(dbits_be_h));
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be_l,
+					 sizeof(dbits_be_l));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:258]: Alignment

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+	case TEE_ALG_SHA384:
+	case TEE_ALG_SHA512:
+		padn = (last < 112) ? (112 - last) : (240 - last);
+
+		rc = ast2600_hace_update(ctx, pad, padn);
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be_h,
+					 sizeof(dbits_be_h));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:253]: Alignment

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+	case TEE_ALG_SHA1:
+	case TEE_ALG_SHA256:
+		padn = (last < 56) ? (56 - last) : (120 - last);
+
+		rc = ast2600_hace_update(ctx, pad, padn);
+		if (rc)
+			return rc;
+
+		rc = ast2600_hace_update(ctx, (uint8_t *)&dbits_be_l,
+					 sizeof(dbits_be_l));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:240]: Alignment

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+		return TEE_ERROR_BUSY;
+
+	mutex_lock(&hace_mtx);
+
+	cache_operation(TEE_CACHEFLUSH, (void *)data, len);
+
+	io_write32(hace_virt + HACE_STS, HACE_STS_HASH_INT);
+
+	io_write32(hace_virt + HACE_HASH_DATA,
+		   virt_to_phys((void *)data));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:103]: Cast not needed?

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+	io_write32(hace_virt + HACE_HASH_DIGEST,
+		   virt_to_phys(hace_ctx->digest));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:105]: Alignment. Or call it `hctx` and everything fits on one line.

**Comment by ChiaweiW** [comment]
Hi @jforissier,

Thanks for your quick feedback.
Could you explain more about the alignment suggestion? 
I was using Linux checkpatch.pl to do the check before the submission.

**Comment by jforissier** [comment]
About statements longer than 80 columns, the Linux coding style specifies "Descendants are always substantially shorter than the parent and are placed substantially to the right.  A very commonly used style is to align descendants to a function open parenthesis.".

So, while the style you used is OK for checkpatch, we prefer to follow the alignment recommendation like so (`▶       ` is a tab character, `·` is a space):

```C
▶       res = some_very_long_function_name(first_parameter, second_parameter,
▶       ▶       ▶       ▶       ▶       ···third_parameter);
```
 
And, you should always fill the 80-column line as much as possible (put as many arguments as possible on the line).

Hope this clarifies.

**Comment by ChiaweiW** [comment]
My local checkpatch.pl seems like be slightly different from that of CI.
I will revise it and re-submit the patch ASAP.

**Comment by jbech-linaro** [comment]
I see IBART timeout on this and other PR's, I'm having a look at it.

**Comment by jbech-linaro** [comment]
> I see IBART timeout on this and other PR's, I'm having a look at it.

For some reason it gets stuck on cloning https://github.com/linaro-swg/patches_hikey. Why that is I don't understand. It works fine when I try here from my own home. I'll continue the investigation.

**Comment by jbech-linaro** [comment]
> > I see IBART timeout on this and other PR's, I'm having a look at it.
> 
> For some reason it gets stuck on cloning https://github.com/linaro-swg/patches_hikey. Why that is I don't understand. It works fine when I try here from my own home. I'll continue the investigation.

It takes ~7-8 minutes to clone everything (even with a mirror). Normally that takes ~50-60 seconds. Could be an issue with the IBART computer, it's pretty old now and maybe the hard-drive is about to give up. Alternatively it could be something with the ISP and the connection, however running speedtest doesn't show signs of that being an issue. For now I've raised the `repo clone` timeout to 10 minutes. I'd like to figure out what's causing this, but I'll do that when I find some spare cycles.

**Comment by ChiaweiW** [comment]
@jbech-linaro Got it! Thanks for taking care of this issue.

File: core/arch/arm/plat-aspeed/platform_ast2600.c [hunk 14]
```diff
@@ -59,6 +59,10 @@ register_phys_mem(MEM_AREA_IO_SEC,
 		  AHBC_BASE,
 		  SMALL_PAGE_SIZE);
 
+register_phys_mem(MEM_AREA_IO_NSEC,
+		  SCU_BASE,
+		  SMALL_PAGE_SIZE);
+
```
Comment (jforissier) [inline core/arch/arm/plat-aspeed/platform_ast2600.c:65]: All those `register_phys_mem()` fit in less than 80 characters. This is a good opportunity to rewrite the whole block as:
```C
register_phys_mem(MEM_AREA_IO_NSEC, CONSOLE_UART_BASE, SMALL_PAGE_SIZE);
register_phys_mem(MEM_AREA_IO_SEC, GIC_BASE + GICD_OFFSET, SMALL_PAGE_SIZE);
register_phys_mem(MEM_AREA_IO_SEC, GIC_BASE + GICC_OFFSET, SMALL_PAGE_SIZE);
register_phys_mem(MEM_AREA_IO_SEC, AHBC_BASE, SMALL_PAGE_SIZE);
register_phys_mem(MEM_AREA_IO_NSEC, SCU_BASE, SMALL_PAGE_SIZE);
```

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+#define   HACE_HASH_CMD_ALG_SHA512	(BIT(6) | BIT(5))
+#define   HACE_HASH_CMD_SHA_BE		BIT(3)
+
+/* buffer size based on SHA-512 need */
+#define HASH_BLK_BUFSZ	128
+#define HASH_DGT_BUFSZ	64
+
+register_phys_mem(MEM_AREA_IO_SEC,
+		  HACE_BASE,
+		  SMALL_PAGE_SIZE);
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:42]: One line

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+
+	do {
+		sts = io_read32(hace_virt + HACE_STS);
+		if (timeout_elapsed(tref)) {
+			rc = TEE_ERROR_TARGET_DEAD;
+			goto unlock_n_out;
+		}
+	} while (!(sts & HACE_STS_HASH_INT));
+
+unlock_n_out:
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:128]: We usually simply call this `out:`.

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+	return TEE_SUCCESS;
+}
+
+static TEE_Result ast2600_hace_update(struct crypto_hash_ctx *ctx,
+				      const uint8_t *data, size_t len)
+{
+	TEE_Result rc = TEE_ERROR_GENERIC;
+	uint32_t left = 0;
+	uint32_t fill = 0;
+	struct ast2600_hace_ctx *hctx = NULL;
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:171]: Suggestion: add `size_t blk_size = hctx->blk_size;` to simplify a bit below.

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+
+static TEE_Result ast2600_hace_final(struct crypto_hash_ctx *ctx,
+				     uint8_t *digest, size_t len)
+{
+	TEE_Result rc = TEE_ERROR_GENERIC;
+	uint32_t last = 0;
+	uint32_t padn = 0;
+	uint8_t pad[HASH_BLK_BUFSZ * 2] = { 0 };
+	uint64_t dbits[2] = { 0, 0 };
+	uint64_t dbits_be[2] = { 0, 0 };
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:216]: `= { };` for all three lines

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+	.free_ctx = ast2600_hace_free,
+	.copy_state = ast2600_hace_copy_state,
+};
+
+static TEE_Result ast2600_hace_alloc(struct crypto_hash_ctx **pctx,
+				     uint32_t algo)
+{
+	struct ast2600_hace_ctx *hctx;
+
+	hctx = calloc(1, sizeof(*hctx));
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:323]:
```suggestion
	struct ast2600_hace_ctx *hctx = calloc(1, sizeof(*hctx));

```

File: core/drivers/crypto/aspeed/hace_ast2600.c [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+
+	*pctx = &hctx->hash_ctx;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ast2600_drvcrypt_register_hash(void)
+{
+	hace_virt = core_mmu_get_va(HACE_BASE, MEM_AREA_IO_SEC,
+				    SMALL_PAGE_SIZE);
```
Comment (jforissier) [inline core/drivers/crypto/aspeed/hace_ast2600.c:369]: Align `SMALL_PAGE_SIZE` under `HACE_BASE`

**Review by jforissier — COMMENTED** [review]
Some more comments. Please squash everything into a single commit, no need for incremental fixes since this PR is small enough I prefer to review everything again. Thanks!

File: core/drivers/crypto/aspeed/hace_ast2600.c (hunk 20 above)
Comment (ChiaweiW) [reply core/drivers/crypto/aspeed/hace_ast2600.c:369]: In VIM with tabstop=8, `SMALL_PAGE_SIZE` seems to be aligned under `HACE_BASE`.
Did I miss something? Thanks.

Comment (ChiaweiW) [reply core/drivers/crypto/aspeed/hace_ast2600.c:369]: The fix commits are squashed for review convenience.

**Comment by jforissier** [comment]
LGTM.

`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by ChiaweiW** [comment]
> LGTM.
> 
> `Acked-by: Jerome Forissier <jerome@forissier.org>`

Tag applied.
Thanks for the help on reviewing the patch and coding suggestions.

**Pull request merged**